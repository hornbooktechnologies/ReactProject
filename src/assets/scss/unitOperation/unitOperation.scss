@import "../_color";
@import "../_variables";

.bg-grey-blue-40-a10 {
  background-color: rgba($grey_blue_40, 0.1);
}
.color-grey-blue-60 {
  color: $grey_blue_60;
}
.min-w-1190 {
  min-width: 1190px;
}
.m-w-110 {
  min-width: 110px;
}
.max-height-500 {
  max-height: $val10 * 50;
}
.font-weight-500 {
  font-weight: 500;
}
// uk Table start
.uk-table {
  &.uk-table-divider {
    th,
    td {
      border: 1px $solid $grey_blue_40;
      text-align: left;
      padding: $val10 $val10 - 2;
      color: $white;
      font-size: $val15 - 1;
      .icn {
        position: $relative;
        top: 2px;
        &.icn-right {
          float: $right;
        }
      }
    }
    th {
      background-color: rgba($white, 0.1);
      font-size: $val15 + 1;
      vertical-align: $middle;
      text-transform: $inherit;
      font-weight: 500;
      text-align: $center;
    }
  }
  &.uk-text-center {
    th,
    td {
      text-align: $center;
    }
  }
  &.uk-text-right {
    th,
    td {
      text-align: $right;
    }
  }
  &.uk-table-sm {
    th {
      font-size: $val15 - 2 !important;
    }
    td {
      font-size: $val15 - 3 !important;
    }
  }
  &.uk-table-middle {
    th,
    td {
      vertical-align: $middle !important;
    }
  }
}
.min-w-86 {
  min-width: 86px;
  width: fit-content;
  margin-top: $val0;
}

@media (max-width: 767px) {
  .media-mt-30 {
    margin-top: 3 * $val10;
  }
}

