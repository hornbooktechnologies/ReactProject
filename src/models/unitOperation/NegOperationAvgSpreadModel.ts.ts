export type NegOperationAvgSpreadModel = {negOperationAvrgSpread:AverageSpreadCardModel[]}
export type NegativeOperationAvgSpreadModel = AverageSpreadCardModel[]

export type AverageSpreadCardModel = {
  title: string,
  Suffix: string,
  plan: AverageSpreadRecordModel
  actual_forcast?: AverageSpreadRecordModel
  yearStartToPresent?: AverageSpreadRecordModel
}

export type AverageSpreadRecordModel = {
  header?:string | undefined
  previousValue?: number | undefined | string
  suffix?: string 
  value?: number | undefined | string
}
