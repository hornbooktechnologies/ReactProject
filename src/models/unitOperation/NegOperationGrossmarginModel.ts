  export type NegGrossmarginTableModel = negativeGrossMargin
  export type NegativeGrossMarginTableModel = negativeGrossMargin
  export type NegativeGrossMarginModel = {negativeGrossMargin:negativeGrossMargin}
  export type negativeGrossMargin = {
    actualOrForcastAnnual:number|null|undefined|string
    actualOrForcastEnd:number|null|undefined|string
    actualOrForcastPresent:number|null|undefined|string
    impactOnEBITDAAnnual:number|null|undefined|string
    impactOnEBITDAEnd:number|null|undefined|string
    impactOnEBITDAPresent:number|null|undefined|string
    planAnnual:number|null|undefined|string
    planEnd:number|null|undefined|string
    planPresent:number|null|undefined|string
    previousActualOrForcastAnnual:number|null|undefined|string
    previousActualOrForcastEnd:number|null|undefined|string
    previousActualOrForcastPresent:number|null|undefined|string
    previousImpactOnEBITDAAnnual:number|null|undefined|string
    previousImpactOnEBITDAEnd:number|null|undefined|string
    previousImpactOnEBITDAPresent:number|null|undefined|string
    previousPlanAnnual:number|null|undefined|string
    previousPlanEnd:number|null|undefined|string
    previousPlanPresent:number|null|undefined|string
    customClassTop: string|undefined
    customClassEnd: string|undefined
    prefix:string|undefined
    suffix:string|undefined
}
