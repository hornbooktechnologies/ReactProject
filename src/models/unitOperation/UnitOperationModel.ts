
export type UnitOperationBaseModel = {
    grossMargin: TimeFrameBaseModel
    amountOfOperatingTime: TimeFrameBaseModel
    averageSpread: TimeFrameBaseModel
    top20DurationOfSpread: DurationOfSpreadTypeModel
    referenceSpecification: ValuesBaseModel
    spreadBarChart:SpreadDurationModel
}

export type Kpi_005ApiRes = {
  PlantID: string
  UnitID: string
  Today: {
    NegativeOperationGrossMargin: NegOperationGrossMargin,
    NegativeOperationTime: NegOperationTime,
    NegativeOperationAvgSpread:NegOperationAvgSpread,
  
  }
  PreviousDay: {
    NegativeOperationGrossMargin: NegOperationGrossMargin,
    NegativeOperationTime: NegOperationTime,
    NegativeOperationAvgSpread:NegOperationAvgSpread,
  }
}
export type NegOperationGrossMargin = {
  Prefix: string
  Suffix: string
  Annual: {
    ImpactOnEBITDA: number,
    Plan: number,
    ActualOrForcast: number,
  },
  YearStartToPresent: {
    ImpactOnEBITDA: number,
    Plan: number,
    ActualOrForcast: number,
  },
  PresentToYearEnd: {
    ImpactOnEBITDA: number,
    Plan: number,
    ActualOrForcast: number,  
  }
}
export type NegOperationTime = {
  Suffix: string
  Annual: {
    Plan: number,
    ActualOrForcast: number,
  },
  YearStartToPresent: {
    Plan: number,
    ActualOrForcast: number,
  },
  PresentToYearEnd: {
    Plan: number,
    ActualOrForcast: number,  
  }
}
export type NegOperationAvgSpread = {
  Suffix: string
  Annual: {
    Plan: number,
    ActualOrForcast: number,
  },
  YearStartToPresent: {
    Plan: number,
    ActualOrForcast: number,
  },
  PresentToYearEnd: {
    Plan: number,
    ActualOrForcast: number,  
  }
}
export type SpreadDurationModel = {
  value?: number;
  Name?: string;
};
export type TimeFrameBaseModel = {
  today: OperatingTimeBaseModel;
  previousDay: OperatingTimeBaseModel;
};

export type OperatingTimeBaseModel = {
  annual: ValuesBaseModel;
  fiscalYearStartToPresent: ValuesBaseModel;
  presentToFiscalYearEnd: ValuesBaseModel;
};

export type ValuesBaseModel = {
    plan?: number
    forecast?: number
    actual?: number
    prefix?: string
    suffix?: string
}

export type DurationOfSpreadTypeModel = {
    actual: DurationOfSpreadModel[]
    forecast: DurationOfSpreadModel[]
}

export type DurationOfSpreadModel = {
    duration: number
    generator: string
    spreadOperationStart: string
    spreadOperationEnd: string
    output: number
    grossMargin: number
    spread: number
}

export type TabDataModel = {
  numberOfCasesTab: string;
  spreadBarData: SpreadBarChartDataModel[];
};

export type SpreadBarChartDataModel = {
    spreadBarData: { Value: number; Name: string }[];
};

export type SpecificationSpreadDataModel = {
  noDataHandler: (val: number, selectValue: string) => void;
  spreadOperation:SpreadOperationOptionBaseDataModel;
  time:number;
};

export interface SpreadOperationOptionBaseDataModel {
  Annual: Object;
  Monthly: Object;
}
export interface SpreadLineChartDataModel {
  data: SpreadLineChartBaseDataModel[] | undefined;
  selectValue: string;
  list?:string[];
  selectedUnitId: string;
  selectedTitleDropdown:string | undefined
}
export interface SpreadLineChartBaseDataModel {
  Period: string;
  Spread: number;
  TimeStamp: number;
  poweTitle: string;
  powerPlantUnitId:string
}
export interface titleDropdownModel {
 value: string;
 label: string;
}
