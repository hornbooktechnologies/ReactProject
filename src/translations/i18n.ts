import { initReactI18next } from 'react-i18next'
import i18n from 'i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import { TRANSLATIONS_EN } from './en/Translations'
import { TRANSLATIONS_JA } from './ja/Translations'

i18n
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    lng: 'en',
    resources: {
      en: {
        translation: TRANSLATIONS_EN,
      },
      ja: {
        translation: TRANSLATIONS_JA,
      },
    },
  })

export default i18n
