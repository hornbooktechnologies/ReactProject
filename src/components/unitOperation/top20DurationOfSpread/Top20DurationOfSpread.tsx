import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { CustomTabModel } from '../../../models/CustomTabModel'
import { DurationOfSpreadTypeModel } from '../../../models/unitOperation/UnitOperationModel'
import { AppConstant } from '../../../utils/AppConstants'
import CustomTabs from '../../common/CustomTabs'
import DurationOfSpreadTable from './DurationOfSpreadTable'

const Top20DurationOfSpread = (top20DurationOfSpread: DurationOfSpreadTypeModel) => {
    const { i18n } = useTranslation()
    const [selectedTab, setSelectedTab] = useState(
        AppConstant.ACTUAL
    );
    
    const handleTabSelection = (selectedTab: string) => {
        setSelectedTab(selectedTab);
    };

    const tabData: CustomTabModel = {
        selectedTab: selectedTab,
        onTabSelect : handleTabSelection,
        tabs: [{
          tabId: 'Actual',
          i8nCode: 'LABELS.ACTUAL'
        }, {
          tabId: 'Forecast',
          i8nCode: 'LABELS.FORECAST'
        }]
    }
    return (
        <>
            <h5 className="mb-24 mt-24 font-18 uk-text-medium">{i18n.t('UNIT_OPERATION.DURATION_OF_SPREAD_HEADER')}</h5>
            <CustomTabs {...tabData} />
            <ul className="uk-switcher mt-24">
              {(() => {
                switch (selectedTab) {
                  case AppConstant.ACTUAL:
                    return <DurationOfSpreadTable {...top20DurationOfSpread.actual} />;
                  case AppConstant.FORECAST:
                    return <DurationOfSpreadTable {...top20DurationOfSpread.forecast} />;
                }
              })()}
            </ul>
        </>
    )
}

export default Top20DurationOfSpread
