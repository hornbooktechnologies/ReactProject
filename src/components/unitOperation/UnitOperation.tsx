import React, { useEffect, useState } from "react";
import { useTranslation } from 'react-i18next';
import { Kpi_005ApiRes, UnitOperationBaseModel } from "../../models/unitOperation/UnitOperationModel";
import { unitOperationApiMockData, unitOperationMockData } from "../../models/unitOperation/_mockData";
import LoadingError from "../common/LoadingError";
import PageHeader from "../common/PageHeader";
import Spinner from "../common/Spinner";
import AmountOfOperatingTime from "./amountOfOperatingTime/AmountOfOperatingTime";
import AverageSpread from "./averageSpread/AverageSpread";
import SpreadNumberOfCases from "./spreadNumberOfCases/SpreadNumberOfCases";
import { spreadOperationMockData } from "../../models/unitOperation/_mockData";
import NegativeOperationGrossMargin from "./NegativeOperationGrossMargin/NegativeGrossMarginTable";
import ReferenceSpecificationOfSpreadOperation from "./specificationSpread/ReferenceSpecificationOfSpreadOperation";
import SpecificationSpread from "./specificationSpread/SpecificationSpread";
import Top20DurationOfSpread from "./top20DurationOfSpread/Top20DurationOfSpread";
import { getCutoffFractionTime } from "../../utils/utils";
import { negOperationAvrgSpreadApiResponse, negOperationGrossmarginApiResponse, negOperationTimeApiResponse } from "../../utils/unitOperation/UnitOperationUtils";
import { NegativeOperationTimeModel} from "../../models/unitOperation/NegOperationTimeModel";
import { NegativeOperationAvgSpreadModel } from "../../models/unitOperation/NegOperationAvgSpreadModel.ts";
import { NegativeGrossMarginTableModel } from "../../models/unitOperation/NegOperationGrossmarginModel";
import "../../assets/scss/annualTarget/_annualstatus.scss";
import '../../assets/scss/unitOperation/unitOperation.scss';
import { DpmService } from "../../services/DpmServices";
import { ApiParamUtils } from "../../utils/apiUtils/ApiParamUtils";

const UnitOperation = () => {
  const { i18n } = useTranslation();
  const [unitOperationData, setUnitOperationData] = useState({} as UnitOperationBaseModel);
  const [negOperationGrossMarginData, setNegOperationGrossMargin] = useState({} as NegativeGrossMarginTableModel)
  const [negOperationTimeData, setNegOperationTime] = useState([] as NegativeOperationTimeModel)
  const [negOperationAvrgSpreadData, setNegOperationAvrgSpread] = useState([] as NegativeOperationAvgSpreadModel)
  const [unitOperationApiData, setUnitOperationApiData] = useState({});
  const [errorThrown, seterrorThrown] = useState(null)
  const [kpiResponse, setKpiResponse] = useState(false)
  const [time, setNoDataDate] = useState(
    getCutoffFractionTime()
  )

  const noDataHandler = (val: number, selectValue: string) => {
    setNoDataDate(val)
  }
    
  useEffect(() => {
    DpmService.fetchKPI_005Data(ApiParamUtils.plantId(),ApiParamUtils.unitCode()) 
    .then((res)=>{
      res=res.data
      if (res.Success) {
        setKpiResponse(res.Success)
        let kpiResponse = res.Data as Kpi_005ApiRes;
        let mappedNegGrossMarginResponse = negOperationGrossmarginApiResponse( kpiResponse);
        let mappedNegOperationTimeResponse = negOperationTimeApiResponse(i18n, kpiResponse);
        let mappedNegAverageSpread = negOperationAvrgSpreadApiResponse(i18n , kpiResponse);
        setNegOperationGrossMargin(mappedNegGrossMarginResponse)
        setNegOperationTime(mappedNegOperationTimeResponse)
        setNegOperationAvrgSpread(mappedNegAverageSpread)
       }
    })
    .catch((error) => {
      setKpiResponse(false)
      seterrorThrown(error)
     })
    setUnitOperationData(unitOperationMockData)
    setUnitOperationApiData(unitOperationApiMockData)
  }, [i18n])

  if (kpiResponse === true && Object.entries(unitOperationData && unitOperationApiData).length > 0) { 
    return (
      <>
        <PageHeader
          title={i18n.t('UNIT_OPERATION.TITLE')}
          description={i18n.t('UNIT_OPERATION.DESCRIPTION')}
        />
        <NegativeOperationGrossMargin negativeGrossMargin = {negOperationGrossMarginData}/>
        <AmountOfOperatingTime negOperationTime = {negOperationTimeData} />
        <AverageSpread negOperationAvrgSpread = {negOperationAvrgSpreadData} />
        <SpecificationSpread spreadOperation={spreadOperationMockData} time = {time} noDataHandler={noDataHandler}/>
        <ReferenceSpecificationOfSpreadOperation {...unitOperationData.referenceSpecification} />
        <SpreadNumberOfCases spreadBarData={unitOperationData.spreadBarChart}/>
        <Top20DurationOfSpread {...unitOperationData.top20DurationOfSpread} />
      </>
    );
  } else if (errorThrown != null) {
    return <LoadingError />;
  } else {
    return <Spinner />;
  }
};
export default UnitOperation;
