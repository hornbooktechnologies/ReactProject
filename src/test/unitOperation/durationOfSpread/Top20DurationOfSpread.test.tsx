import React from "react";
import renderer from "react-test-renderer";
import i18n from "../../i18nForTests";
import { I18nextProvider } from "react-i18next";
import { DurationOfSpreadTypeModel } from "../../../models/unitOperation/UnitOperationModel";
import Top20DurationOfSpread from "../../../components/unitOperation/top20DurationOfSpread/Top20DurationOfSpread";
describe("Component: Top20DurationOfSpread", () => {
  it("Top20DurationOfSpread renders correctly", () => {
    const mockData: DurationOfSpreadTypeModel = {
      actual: [
        {
          duration: 23,
          generator: "Generator 5",
          grossMargin: -1.01,
          output: 478,
          spread: -1.1,
          spreadOperationEnd: "2022-03-11T12:07:33.792Z",
          spreadOperationStart: "2022-03-11T12:07:33.792Z",
        },
      ],
      forecast:[
          {
            duration: 23,
            generator: "Generator 2",
            grossMargin: -1.01,
            output: 448,
            spread: -1.1,
            spreadOperationEnd: "2022-03-11T12:07:33.793Z",
            spreadOperationStart: "2022-03-11T12:07:33.793Z",
          }
      ]
    };
    const tree = renderer
      .create(
        <I18nextProvider i18n={i18n}>
          <Top20DurationOfSpread {...mockData} />
        </I18nextProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
