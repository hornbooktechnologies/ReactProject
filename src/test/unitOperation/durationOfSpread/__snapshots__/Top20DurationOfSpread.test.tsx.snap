// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Top20DurationOfSpread Top20DurationOfSpread renders correctly 1`] = `
Array [
  <h5
    className="mb-24 mt-24 font-18 uk-text-medium"
  >
    UNIT_OPERATION.DURATION_OF_SPREAD_HEADER
  </h5>,
  <ul
    className="uk-tab "
    uk-tab="true"
  >
    <li
      className="uk-active"
      onClick={[Function]}
    >
      <a
        href="about:blank"
      >
         
        LABELS.ACTUAL
      </a>
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <a
        href="about:blank"
      >
         
        LABELS.FORECAST
      </a>
    </li>
  </ul>,
  <ul
    className="uk-switcher mt-24"
  >
    <li
      className="uk-active"
    >
      <div
        className="mt-24"
      >
        <div
          className="uk-overflow-auto max-height-500"
        >
          <table
            className="uk-table uk-table-divider uk-text-right min-w-1190"
          >
            <thead
              className="spreadOperationTable"
            >
              <tr>
                <th
                  className="uk-text-center"
                >
                  UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.NO
                </th>
                <th
                  className="uk-text-center"
                >
                  UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.DURATION
                </th>
                <th
                  className="uk-text-center"
                >
                  UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.OPERATION_START
                </th>
                <th
                  className="uk-text-center"
                >
                  UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.OPERATION_END
                </th>
                <th
                  className="uk-text-center"
                >
                  <span
                    className="uk-display-inline-block"
                  >
                    UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.OUTPUT
                    <span
                      className="font-12"
                    >
                      (
                      UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.AVERAGE
                      )
                    </span>
                    <span
                      className="uk-text-right font-12 uk-display-block"
                    >
                      【
                      UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.OUTPUT_UNIT
                      】
                    </span>
                  </span>
                </th>
                <th
                  className="uk-text-center"
                >
                  UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.GROSS_MARGIN
                  <span
                    className="uk-display-inline-block"
                  >
                    <span
                      className="uk-text-right font-12 uk-display-block"
                    >
                      【
                      UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.GROSS_MARGIN_UNIT
                      】
                    </span>
                  </span>
                </th>
                <th
                  className="uk-text-center"
                >
                  <span
                    className="uk-display-inline-block"
                  >
                    UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.SPREAD
                    <span
                      className="font-12"
                    >
                      (
                      UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.AVERAGE
                      )
                    </span>
                    <span
                      className="uk-text-right font-12 uk-display-block"
                    >
                      【
                      UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.SPREAD_UNIT
                      】
                    </span>
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  1
                </td>
                <td>
                  23
                </td>
                <td
                  className="uk-text-center"
                >
                  2022/3/11 17:37
                </td>
                <td
                  className="uk-text-center"
                >
                  2022/3/11 17:37
                </td>
                <td>
                  478
                </td>
                <td>
                  -1.01
                </td>
                <td>
                  -1.10
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </li>
  </ul>,
]
`;
